### ISSUE TITLE
poetry add crash when using token for private git package: urllib3 is not defined

### ISSUE THREAD

**Issue by ekaj2** [python-poetry/poetry#8526]
- **Poetry version**: 1.6.1
- **Python version**: 3.11.5
- **OS version and name**: macOS 13.5.2
- **pyproject.toml**: not relevant

- [X] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) stable Poetry version, installed using a recommended method.
- [X] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [X] I have consulted the [FAQ](https://python-poetry.org/docs/faq/) and [blog](https://python-poetry.org/blog/) for any relevant entries or release notes.
- [X] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option) and have included the output below.

## Issue

`poetry add` fails when using a GitHub private access token. There's clearly a few issues here, but the logged output is pretty self-explanatory. I'm happy to promptly answer any questions though.

```bash
dj-proj ❯ poetry --verbose add "git+https://github_pat_REDACTED:x-oauth-basic@github.com/myorg/myproject.git"
Using virtualenv: /Users/myname/Library/Caches/pypoetry/virtualenvs/myappname-py3.11

  UnicodeError

  encoding with 'idna' codec failed (UnicodeError: label too long)

  at ~/Library/Application Support/pypoetry/venv/lib/python3.11/site-packages/urllib3/util/connection.py:56 in create_connection
       52│     # The original create_connection function always returns all records.
       53│     family = allowed_gai_family()
       54│
       55│     try:
    →  56│         host.encode("idna")
       57│     except UnicodeError:
       58│         raise LocationParseError(f"'{host}', label empty or too long") from None
       59│
       60│     for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):

The following error occurred when trying to handle this error:


  LocationParseError

  Failed to parse: 'github_pat_REDACTED', label empty or too long

  at ~/Library/Application Support/pypoetry/venv/lib/python3.11/site-packages/urllib3/util/connection.py:58 in create_connection
       54│
       55│     try:
       56│         host.encode("idna")
       57│     except UnicodeError:
    →  58│         raise LocationParseError(f"'{host}', label empty or too long") from None
       59│
       60│     for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
       61│         af, socktype, proto, canonname, sa = res
       62│         sock = None

The following error occurred when trying to handle this error:


  NameError

  name 'urllib3' is not defined

  at ~/Library/Application Support/pypoetry/venv/lib/python3.11/site-packages/dulwich/client.py:2241 in _http_request
      2237│             else:
      2238│                 resp = self.pool_manager.request(
      2239│                     "POST", url, headers=req_headers, body=data, preload_content=False
      2240│                 )
    → 2241│         except urllib3.exceptions.HTTPError as e:
      2242│             raise GitProtocolError(str(e)) from e
      2243│
      2244│         if resp.status == 404:
      2245│             raise NotGitRepository()
```

**Comment by dimbleby** [comment]
None of this is in code that lives in this repository.  You're probably in the wrong place; or anyway need to show why why you think these errors can be fixed here

**Comment by totoroot** [comment]
I just wanted to report that I also ran into this issue when trying to add a private repo from GitLab to my project with `poetry add` by using an API_TOKEN:
`poetry --verbose add git+https://oauth2:<redacted>@gitlab.com/<organisation>/<repository>.git`

- Poetry version: 1.6.1
- Python version: 3.10.13
- OS version and name: NixOS 23.11.20231016.ca012a0 (Tapir) x86_64

Also, I think that, although the code that causes it might not live in this repository, it is a valid issue.
If a user of poetry runs into this and is not sure how to resolve it, they might look through the issues of poetry, since the error came up when executing the `poetry add` command.

So if we can find a solution to this, even if it isn't by means of a code change in this repo, it helps this user of poetry.

@dimbleby Don't you agree?

**Comment by dimbleby** [comment]
of course fixing poetry dependencies helps poetry: but if you want to make that happen then you should go directly to those dependencies.

eg the failed urllib3 import bug is clearly a dulwich problem and wants reporting there

**Comment by totoroot** [comment]
Is it just a dulwich problem though?

To find out if it was, I also tested setting `experimental.system-git-client` to `true`, as that should not use dulwich as far as I'm aware. It looks like there is some something wrong with the parsing of the repository's URL if you pass a token.

For example I can clone a private repository using git just fine:

```sh
git clone https://oauth2:<repo-access-token>@gitlab.com/<organisation>/<repository>.git
```
The `oauth2:<repo-access-token>` is how you have to pass the token if you want to access a private repo on GitLab via HTTPS.
If you look at @ekaj2's initial issue description, it is a similar situation for GitHub.

When I now try to add the private repo using the token with `poetry -v add git+https://oauth2:<repo-access-token>@gitlab.com/<organisation>/<repository>.git`, I get the following error:

```py
Using virtualenv: .../.venv

  CalledProcessError

  Command '['git', 'clone', '--recurse-submodules', '--', 'https://oauth2/<repo-access-token>@gitlab.com/<organisation>/<repository>.git', '.../.venv/src/...']' returned non-zero exit status 128.

  at /nix/store/vagb0sjv83ybi435i6iiv10hjrdghph9-python3-3.10.12/lib/python3.10/subprocess.py:526 in run
       522│             # We don't call process.wait() as .__exit__ does that for us.
       523│             raise
       524│         retcode = process.poll()
       525│         if check and retcode:
    →  526│             raise CalledProcessError(retcode, process.args,
       527│                                      output=stdout, stderr=stderr)
       528│     return CompletedProcess(process.args, retcode, stdout, stderr)
       529│
       530│

The following error occurred when trying to handle this error:


  PoetryConsoleError

  Failed to clone https://oauth2/<repo-access-token>@gitlab.com/<organisation>/<repository>.git, check your git configuration and permissions for this repository.

  at /nix/store/g2zlcv3829f80m9axwbzh0ar1jsvl220-python3.10-poetry-1.6.1/lib/python3.10/site-packages/poetry/vcs/git/backend.py:227 in _clone_legacy
      223│
      224│         try:
      225│             SystemGit.clone(url, target)
      226│         except CalledProcessError:
    → 227│             raise PoetryConsoleError(
      228│                 f"Failed to clone {url}, check your git configuration and permissions"
      229│                 " for this repository."
      230│             )
      231│
```

If you look at the clone command, the `:` between `oauth2` and the token got replaced with a `/` .

Wouldn't that have happened somewhere in poetry?

**Comment by ekaj2** [comment]
Sorry @dimbleby, I didn't get notifications until @totoroot mentioned me. To be fair, I had to look up what dulwich was - I had hacked this dependency setup one night based on several things I read online. I work at a startup, so "tmw" deadlines happen almost constantly 

The reason I jumped to say this is a poetry issue is because I'd used the exact same dependency with both virtualenv directly as well as the Heroku build system which both work fine with it. I just switched to poetry because our requirements were getting out of hand, so when it didn't work, I assumed it was poetry.

Can't dig into it just yet, but hopefully will have some time over the weekend

**Comment by dimbleby** [comment]
@totoroot 
```
  NameError

  name 'urllib3' is not defined
```
is absolutely clearly a dulwich problem and wants reporting to that repository

poetry may very well have other bugs but do check for duplicates before raising new reports, I'm sure I've seen discussion of gitlab and access tokens here previously.

**Comment by danny-vayu** [comment]
> @totoroot
> 
> ```
>   NameError
> 
>   name 'urllib3' is not defined
> ```
> 
> is absolutely clearly a dulwich problem and wants reporting to that repository
> 
> poetry may very well have other bugs but do check for duplicates before raising new reports, I'm sure I've seen discussion of gitlab and access tokens here previously.

This response is puzzling: I don't care about dulwich or any other dependencies Poetry uses. I am a user of Poetry - and from my perspective it crashes.  

If I build software that depends on library X, and library X throws, I don't tell my users: "it's not my problem. It's a bug in X".

**Comment by dimbleby** [comment]
haha, you mistake a github issue for paid support.  This is a do-it-yourself world!

Happily there are others out there who know how the game is played: this was already reported to dulwich at https://github.com/jelmer/dulwich/issues/1208, and fixed in https://github.com/jelmer/dulwich/pull/1209

this can be closed

**Comment by keikoro** [comment]
To add to what @danny-vayu 
> This response is puzzling: I don't care about dulwich or any other dependencies Poetry uses. I am a user of Poetry - and from my perspective it crashes.

and @totoroot have said (this is basically why I'm here now):
> Also, I think that, although the code that causes it might not live in this repository, it is a valid issue.
If a user of poetry runs into this and is not sure how to resolve it, they might look through the issues of poetry, since the error came up when executing the `poetry add` command.

I haven't experienced OP's particular issue but I've run into the same `name 'urllib3' is not defined` error several times recently.

Reason? I was offline, but unaware of it (got disconnected). And it didn't occur to me that that could be the problem because I was running `poetry lock --no-update` while experimenting with different dependency configs, packages I'd already installed/removed previously, so expected to be part of the cache.

Now, in my case, of course the problem could be that I'm misunderstanding how the cache works. (A topic about which I couldn't find much on the poetry website, nor elsewhere online.) Maybe `poetry lock` _always_ tries to connect to PyPI or Git sources, without exception. ... But me not being able to infer that or easily read up on/learn about this just points to the same problem (of poor UX).

**Comment by ekaj2** [comment]
Haven't had a chance to dive into this more because I'm at a startup where everything is a stressfully fast pace rn. The reason I raised the issue is this was a migration from using pip to poetry. All of these deps resolve correctly when using pip, but broke with poetry. I was under the impression that poetry is a drop-in replacement and would be expected to work here.

Feel free to close if it's out of the scope of this lib.

**Comment by dimbleby** [comment]
the dulwich fix has just been released, even more than ever this can be closed

**Comment by mmetc** [comment]
This issue has been fixed in dulwich 0.21.7, but poetry is still using 0.21.2.

**Comment by dimbleby** [comment]
Poetry requirements allow recent dulwich, if you want to upgrade it you can
